designs/replicated-pv/mayastor/rwx-kubevirt.md

```diff
@@ -0,0 +1,132 @@
+- Support Persistent Reservations (optional) for fencing.
+
+### 🚫 Non-Goals
+
+- RWX support for filesystem volumes.
+- RWX support for non-KubeVirt workloads.
+- VM-level fencing or quorum management (we don't want to share, but rather to switch!).
+
+---
+
+## 🧵 Proposal
+
+As per my understanding, KubeVirt itself handles the flushing and switching of the storage to the new node, which means we don't even have to implement reservations.
+The CSI spec allows for many access modes, example:
+
+- ***MULTI_NODE_SINGLE_WRITER***
+  A volume can be published at multiple nodes simultaneously. Only one of the node can be used as read/write. The rest will be readonly.
+- ***MULTI_NODE_MULTI_WRITER***
+  A volume can be published as read/write at multiple nodes simultaneously.
+
+However, KubeVirt is consuming K8s volumes, and there's no such thing as `ReadWriteOneReadMany`, only `ReadWriteMany`. As such, we have to support the `ReadWriteMany` for Block volumes by implementing the `MultiNodeMultiWriter` capability.
+
+Mayastor currently supports only `NVMe-Of` at the top, and this is suitable for exposing a volume to multiple nodes at a time.
+In addition, the `nvme` protocol itself does support reservations as a mechanism for:
+
+- Controlling access to shared namespaces
+- Preventing conflicts between hosts
+- Ensuring data integrity and consistency
+
+Considering this OEP is targeting KubeVirt specifically, then we may skip implementing access control through the reservations, on the premise that KubeVirt is already taking care of this by itself.
+
+### 📚 User Stories
+
+#### 🧑‍💻 Story 1
+
+As a user, I want to live-migrate my KubeVirt VMs from one node to another. I accept that I must use block volumes in order to achieve this.
+
+### 🛠️ Implementation Details/Notes/Constraints
+
+Following a top-down approach, here are the required modifications:
+
+1. Extend Mayastor CSI to advertise `MultiNodeMultiWriter`, but only for block mode volumes.
+   1. this means we'd reject it for fs volumes, since that cannot possibly work!
+2. Core agent must now accept publishes, even if the volume is already published, but for a different node
+3. Core agent must collate all valid initiator nqns when building the access control list for the NVMe-oF target.
+
+<br>
+
+Here's a depiction on how the connected nodes would look like:
+
+```mermaid
+graph LR;
+    subgraph node-1
+        kubelet-1
+        csi-node-1
+        nvme-of-initiator-1
+    end
+    subgraph node-2
+        kubelet-2
+        csi-node-2
+        nvme-of-initiator-2
+    end
+    subgraph node-3
+        io-engine-3
+    end
+
+    subgraph ControlPlane
+        subgraph Agents
+            Core
+        end
+        Rest
+    end
+
+    csi-controller
+
+    kubelet-1 --> |CSI Api| csi-node-1
+    kubelet-2 --> |CSI Api| csi-node-2
+    csi-controller --> |Public Api| Rest
+    csi-node-1 --> |Public Api| Rest
+    csi-node-2 --> |Public Api| Rest
```

Comment:
Why does csi node need to interact with rest api? What information are we gathering here?

Reply:
Was only here to show we have a connection, but actually doesn't need to, only confusing things... removed it